tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp

```diff
@@ -2301,21 +2380,54 @@ void SwiftLangSupport::findRelatedIdentifiersInFile(
         if (VD->isOperator())
           return;
 
-        RelatedIdScanner Scanner(SrcFile, BufferID, VD, Ranges);
+        // Record ranges in a set first so we don't record some ranges twice.
+        // This could happen in capture lists where e.g. `[foo]` is both the
+        // reference of the captured variable and the declaration of the
+        // variable usable in the closure.
+        llvm::SmallDenseSet<std::pair<unsigned, unsigned>, 8> RangesSet;
+
+        // List of decls whose ranges should be reported as related identifiers.
+        SmallVector<ValueDecl *, 2> Worklist;
+        Worklist.push_back(VD);
+
+        // Decls that we have already visited, so we don't walk circles.
+        SmallPtrSet<ValueDecl *, 2> VisitedDecls;
```

Comment:
LLVM has `SmallSetVector` which you could use instead

Reply:
Why do you think `SmallSetVector` would be better than a `SmallPtrSet` here? I don’t think we need any ordering guarantees here.

Reply:
The idea would be to merge `Worklist` and `VisitedDecls`.

Reply:
Ah, sorry, SetVector also removes from the `Set` in a `pop_back`. Ignore me.